// express-vue-renderer/lib/models/defaults.js
'use strict';

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

var path = require('path');
var deepmerge = require('deepmerge');
var Layout = require('./layout');

function concatMerge(destinationArray, sourceArray) {
    var finalArray = destinationArray.concat(sourceArray);
    //Dedupes dupes... obviously... but theres a problem here
    // return dedupe(finalArray);
    return finalArray;
}

var Defaults = function () {
    function Defaults() {
        var options = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};

        _classCallCheck(this, Defaults);

        this.options = options;
        this.layout = new Layout.Layout(options.layout);

        if (options.style) {
            this.style = options.style;
        }

        if (options.rootPath) {
            this.rootPath = path.resolve(options.rootPath);
        }
        if (options.component) {
            this.component = options.component;
        }

        if (options.vue) {
            this.vue = options.vue;
        } else {
            this.vue = {};
        }
        if (options.data) {
            this.data = options.data;
        } else {
            this.data = {};
        }
    }

    _createClass(Defaults, null, [{
        key: 'mergeObjects',
        value: function mergeObjects(globalObject, newObject) {
            var mergedObject = deepmerge(globalObject, newObject, { arrayMerge: concatMerge });
            return mergedObject;
        }
    }]);

    return Defaults;
}();

module.exports = Defaults;
