// express-vue-renderer/lib/renderer/render.js
'use strict';

var Utils = require('../utils');
var Vue = require('vue');

var butternut = require('butternut');
var buttterNutOptions = {
    sourceMap: false
};

function createApp(script) {
    return new Vue(script);
}

function layoutUtil(component) {
    var layout = {};
    layout.style = '';
    layout.script = component.script;
    if (component.style.length > 0) {
        layout.style += component.style;
    }
    return layout;
}

function renderedScript(script, router) {
    var routerString = router !== undefined ? 'const __router = new VueRouter(' + Utils.scriptToString(router) + ');' : '';
    var scriptString = Utils.scriptToString(script);
    var debugToolsString = '';

    if (router !== undefined) {
        scriptString = scriptString.substr(0, 1) + 'router: __router,' + scriptString.substr(1);
    }
    if (process.env.VUE_DEV) {
        debugToolsString = 'Vue.config.devtools = true;';
    }
    var javaScriptString = '(function () {\'use strict\';' + routerString + 'var createApp = function () {return new Vue(' + scriptString + ')};if (typeof module !== \'undefined\' && module.exports) {module.exports = createApp} else {this.app = createApp()}}).call(this);' + debugToolsString + 'app.$mount(\'#app\');';
    var finalString = butternut.squash(javaScriptString, buttterNutOptions).code;

    return '<script>' + finalString + '</script>';
}

function renderHtmlUtil(component) {
    var layout = layoutUtil(component);
    var renderedScriptString = renderedScript(layout.script);
    var app = createApp(layout.script);

    return {
        app: app,
        scriptString: renderedScriptString,
        layout: layout,
        scriptStringRaw: Utils.scriptToString(layout.script)
    };
}

module.exports.layoutUtil = layoutUtil;
module.exports.renderHtmlUtil = renderHtmlUtil;
